### PR TITLE
Light by light modifier 1100 pre1

### PR DESCRIPTION
**Pull request by rchudasa**
#### PR description:
This PR is submitted to reconstruct low pT photons down to 2 GeV for light by light scattering. It is just about relaxing few thresholds in python config with no changes in c++ source code.  A modifier has been added to access the low pt photon reconstruction. Photon and electron efficiency plots and timing with default and modified reconstruction can be found in slides [1] presented in Reconstruction and Analysis Tools meeting. The modified reconstruction has been reviewed by Egamma POG convener @Sam-Harper .
[1]https://indico.cern.ch/event/836517/contributions/3513898/attachments/1887027/3111085/cms_reco_meeting_26thJuly_2019.pdf

### CODE REVIEW

File: Configuration/Eras/python/Modifier_lightByLightLowPt_cff.py [hunk 1]
```diff
@@ -0,0 +1,4 @@
+import FWCore.ParameterSet.Config as cms
```
Comment (slava77) [inline Configuration/Eras/python/Modifier_lightByLightLowPt_cff.py]: Configuration/ProcessModifiers/python might be a better place

**Comment by slava77** [comment]
> Photon and electron efficiency plots and timing with default and modified reconstruction can be found in slides [1] presented in Reconstruction and Analysis Tools meeting. The modified reconstruction has been reviewed by Egamma POG convener @Sam-Harper .
> [1]https://indico.cern.ch/event/836517/contributions/3513898/attachments/1887027/3111085/cms_reco_meeting_26thJuly_2019.pdf

I missed this part of the meeting.
What do you imply by the "default reco"?
Naively, given the 10_3_X context and HI PDs, I can guess that this is about pp_on_AA,
but I'm not sure.

Is this expected to be a once-in-CMS-lifetime for 10_3_X use case or should this be a supported workflow?

**Comment by rchudasa** [comment]
> > Photon and electron efficiency plots and timing with default and modified reconstruction can be found in slides [1] presented in Reconstruction and Analysis Tools meeting. The modified reconstruction has been reviewed by Egamma POG convener @Sam-Harper .
> > [1]https://indico.cern.ch/event/836517/contributions/3513898/attachments/1887027/3111085/cms_reco_meeting_26thJuly_2019.pdf
> 
> I missed this part of the meeting.
> What do you imply by the "default reco"?
> Naively, given the 10_3_X context and HI PDs, I can guess that this is about pp_on_AA,
> but I'm not sure.
> 
> Is this expected to be a once-in-CMS-lifetime for 10_3_X use case or should this be a supported workflow?

The default reco is pp reconstruction with era "Run2_2018" and on top of that light by light modifier has been added. We don't use or modify pp_on_AA era as it modifies many Egamma thresholds to higher values which we don't want. 
This is a supported workflow for higher CMSSW release and even for pp analyses, if required. :)

**Comment by slava77** [comment]
> The default reco is pp reconstruction with era "Run2_2018" 

uhm, so this make a stronger case for backporting this at best to 10_6_X to be based on the UL release.

**Comment by slava77** [comment]
@rchudasa 
there is a merge conflict in this PR now
Please update.

**Review by rchudasa — COMMENTED** [review]
> @rchudasa
> there is a merge conflict in this PR now
> Please update.

@slava77 , I just committed a change now. Please note that I have not changed the location of lightbylight modifier to ProcessModifier repo yet, as cmsDriver command could not identify that.

**Comment by slava77** [comment]
On 8/1/19 8:34 AM, rchudasa wrote:
> Please note that I have not changed the location of lightbylight
> modifier to ProcessModifier repo yet, as cmsDriver command could not
> identify that.
> 
> ------------------------------------------------------------------------

it should be passed with --procModifiers options

**Comment by slava77** [comment]
>  I tried to keep this modifier in ProcessModifiers repo as suggested by you and tried to run cmsDriver command with --era Run2_2018 --procModifiers lightByLightLowPt but it tries to find this Modifier in Eras repo, could you please suggest what else should I modify, so that cmsDriver command recognizes it from ProcessModifiers repo?

This sounds strange. By construction ```--procModifiers``` looks in Configuration.ProcessModifiers.
https://github.com/cms-sw/cmssw/blob/master/Configuration/Applications/python/ConfigBuilder.py#L2092
Note that you should drop ```Modifier_``` prefix from the file name.

**Comment by rchudasa** [comment]
> > I tried to keep this modifier in ProcessModifiers repo as suggested by you and tried to run cmsDriver command with --era Run2_2018 --procModifiers lightByLightLowPt but it tries to find this Modifier in Eras repo, could you please suggest what else should I modify, so that cmsDriver command recognizes it from ProcessModifiers repo?
> 
> This sounds strange. By construction `--procModifiers` looks in Configuration.ProcessModifiers.
> https://github.com/cms-sw/cmssw/blob/master/Configuration/Applications/python/ConfigBuilder.py#L2092
> Note that you should drop `Modifier_` prefix from the file name.

Yes, I dropped Modifier_ from the name but the problem is, the cmsDriver command tries to search in Eras repository, although I gave --procModifiers option. Here's the error: CMSSW_11_0_X_2019-07-29-2300/python/Configuration/StandardSequences/Eras.py", line 81, in __init__
    eObj=getattr(__import__('Configuration.Eras.Modifier_'+e+'_cff',globals(),locals(),[e],0),e)
ImportError: No module named Modifier_lightByLightLowPt_cff

**Comment by makortel** [comment]
> Yes, I dropped Modifier_ from the name but the problem is, the cmsDriver command tries to search in Eras repository, although I gave --procModifiers option. Here's the error: CMSSW_11_0_X_2019-07-29-2300/python/Configuration/StandardSequences/Eras.py", line 81, in init
> eObj=getattr(import('Configuration.Eras.Modifier_'+e+'_cff',globals(),locals(),[e],0),e)
> ImportError: No module named Modifier_lightByLightLowPt_cff
> 

You need to remove your additions  to `Eras.py`, namely
https://github.com/cms-sw/cmssw/blob/645b6198df584667c0b6265af657135285c26f1f/Configuration/StandardSequences/python/Eras.py#L70

**Comment by slava77** [comment]
about the naming of the modifier, I'd like to propose some decoration that refers to a POG or a PAG. One somewhat clear option is ```egamma_photon_lowPt```.
The specific goal of the proposed change is really to capture the (semi)exclusive diphoton production. Perhaps in the HI case gamma-gamma dominates, but at lower Z the diffractive component can be large as well.

OTOH, if the longer term goal for this is to include coverage for other final states, perhaps a less specific name is more practical like ```exclusive_lowPt```

**Comment by rchudasa** [comment]
> about the naming of the modifier, I'd like to propose some decoration that refers to a POG or a PAG. One somewhat clear option is `egamma_photon_lowPt`.
> The specific goal of the proposed change is really to capture the (semi)exclusive diphoton production. Perhaps in the HI case gamma-gamma dominates, but at lower Z the diffractive component can be large as well.
> 
> OTOH, if the longer term goal for this is to include coverage for other final states, perhaps a less specific name is more practical like `exclusive_lowPt`

@slava77 , currently we don't have any plan to extend this to any other final state and the current reconstruction modifies Photons and electrons both and it's more for exclusive process, we would like to propose 'egamma_lowPt_exclusive'. Is it ok with you?

**Comment by slava77** [comment]
On 8/2/19 9:24 AM, rchudasa wrote:
> we would like to propose 'egamma_lowPt_exclusive'. Is it ok with you?

sounds fine to me

**Comment by fabiocos** [comment]
@rchudasa I agree with comments by @slava77 both about the modifier location (we are already abusing of Era, and a cleaning would be advisable https://github.com/cms-sw/cmssw/issues/22718 ) and the name, I would prefer to avoid overspecializing the name of top level configurations

**Comment by rchudasa** [comment]
@slava77 @fabiocos @makortel : Thank you for your suggestions. I have renamed the light by light modifier to egamma_lowPt_exclusive and also kept it in ProcessModifiers repository.

**Comment by fabiocos** [comment]
please test

not really testing the new feature here

**Comment by slava77** [comment]
@rchudasa @echapon 
please post a cmsDriver command which can be used to test the proposed update together with an appropriate input file (preferably readable via xrootd).
Thank you.

File: RecoEgamma/EgammaPhotonProducers/python/photons_cfi.py [hunk 2]
```diff
@@ -150,3 +150,11 @@
...
     RecHitSeverityToBeExcludedEE = cleanedHybridSuperClusters.RecHitSeverityToBeExcluded,
 )
+
+from Configuration.ProcessModifiers.egamma_lowPt_exclusive_cff import egamma_lowPt_exclusive
+egamma_lowPt_exclusive.toModify(photons,
+                           minSCEtBarrel = 1.0, #default 10
+                           minSCEtEndcap = 1.0) #default 10
+egamma_lowPt_exclusive.toModify(islandPhotons,
+                           minSCEtBarrel = 1.0, #default 5
+                           minSCEtEndcap = 1.0) #default 15
```
Comment (slava77) [inline RecoEgamma/EgammaPhotonProducers/python/photons_cfi.py]: this is not a part of the regular pp reco.
Why is this modified?

File: RecoParticleFlow/PFProducer/python/particleFlowBlock_cfi.py [hunk 3]
```diff
@@ -209,3 +209,9 @@ def _findIndicesByModule(name):
     particleFlowBlock,
     elementImporters = _addTimingLayer
 )
+
+from Configuration.ProcessModifiers.egamma_lowPt_exclusive_cff import egamma_lowPt_exclusive
+egamma_lowPt_exclusive.toModify(particleFlowBlock.elementImporters[2],
```
Comment (slava77) [inline RecoParticleFlow/PFProducer/python/particleFlowBlock_cfi.py]: please use ```_findIndicesByModule``` instead of direct indexing.
A less preferred solution would be to move this change up, to be right after the definition of particleFlowBlock, before all other modifiers are called

**Comment by rchudasa** [comment]
> @rchudasa @echapon
> please post a cmsDriver command which can be used to test the proposed update together with an appropriate input file (preferably readable via xrootd).
> Thank you.
@slava77 @echapon 
Here's the cmsDriver command with xrootd input 

cmsDriver.py step3 --mc --eventcontent AODSIM --datatier AODSIM --conditions 110X_upgrade2018_realistic_Candidate_2019_08_03_11_37_17 --step RAW2DIGI,L1Reco,RECO --geometry DB:Extended -n -1 --era Run2_2018 --procModifiers egamma_lowPt_exclusive --customise Configuration/DataProcessing/Utils.addMonitoring --filein=root://cmsxrootd.fnal.gov//store/group/phys_diffraction/lbyl_2018/for_pr/FlatGamma_pt_digi_raw_11X.root --customise_commands 'process.AODSIMoutput.outputCommands.extend(["keep *_towerMaker_*_*", "keep *_siPixelRecHits_*_*", "keep *_siPixelClusters_*_*", "keep *_interestingGedEgammaIsoESDetId_*_*", "keep *_interestingGedEgammaIsoHCALDetId_*_*", "keep *_interestingGedEleIsoDetIdEB_*_*", "keep *_interestingGedEleIsoDetIdEE_*_*", "keep *_interestingGedGamIsoDetIdEB_*_*","keep *_interestingGedGamIsoDetIdEE_*_*", "keep EBDigiCollection_ecalDigis_*_*","keep EEDigiCollection_ecalDigis_*_*","keep ESDigiCollection_ecalPreshowerDigis_*_*","keep ZDCRecHitsSorted_zdcreco_*_*"])'

**Comment by slava77** [comment]
@rchudasa 
Thank you for the cmsDriver command.
if you need modifications to the output AODSIM, it is more appropriate to make them with egamma_lowPt_exclusive in the code base. Please add them to this PR.
Please also escape the command in full with \`\`\` on both sides so that it's displayed appropriately (https://github.com/cms-sw/cmssw/pull/27633#issuecomment-519177827 looks like missing some underscores and stars due to confusion with the markdown special characters)

Please update your tests (extending what was shown in the reco meeting) with details of the output file sizes. I expect that the extra data in your ```keep ``` statements increases the AOD size by a factor of a few.

**Comment by slava77** [comment]
how was the file FlatGamma_pt_digi_raw_11X.root (proposed for testing)  made?

**Comment by slava77** [comment]
> how was the file FlatGamma_pt_digi_raw_11X.root (proposed for testing) made?

Looking at the provenance, I see that this is a photon gun MC sample with photon in |eta|<3 and pt from 1 to 10 GeV.
I think that this is insufficient for technical performance tests.
Please also provide a data file.

**Comment by rchudasa** [comment]
> > how was the file FlatGamma_pt_digi_raw_11X.root (proposed for testing) made?
> 
> Looking at the provenance, I see that this is a photon gun MC sample with photon in |eta|<3 and pt from 1 to 10 GeV.
> I think that this is insufficient for technical performance tests.
> Please also provide a data file.
Yes, the digi raw file was produced with flat pt photon sample produced for pt: 1-20 GeV within |eta|<3. The dtata file can be found here: /eos/cms/store/group/phys_diffraction/lbyl_2018/F9B1A876-3F7D-8449-8437-894248B60877.root. This data file contains good lumi sections.

File: RecoEgamma/EgammaPhotonProducers/python/photons_cfi.py (hunk 2 above)
Comment (rchudasa) [reply RecoEgamma/EgammaPhotonProducers/python/photons_cfi.py]: ok, we will not change the island Photon thresholds (updating it in the next commit).

**Comment by slava77** [comment]
> The dtata file can be found here: /eos/cms/store/group/phys_diffraction/lbyl_2018/F9B1A876-3F7D-8449-8437-894248B60877.root. This data file contains good lumi sections.

Which GT should be used with this data file? Is the run2_data OK or do we need run2_data_promptlike_hi ?
Please clarify, thank you.

**Comment by slava77** [comment]
> Which GT should be used with this data file? Is the run2_data OK or do we need run2_data_promptlike_hi ?

I see that it runs with auto:run2_data. perhaps this is good enough for the tests.

**Comment by slava77** [comment]
> Please update your tests (extending what was shown in the reco meeting) with details of the output file sizes. I expect that the extra data in your `keep ` statements increases the AOD size by a factor of a few.

I should have been looking at the slides more carefully.
There is a "Size:" bullet for both cases.
However, please clarify if the default reco size was measured with the keep statements or using the default AOD definition.

**Comment by slava77** [comment]
about timing: I was not able to reproduce the significant increase in CPU time between the default and the modified recos (you showed the avg go up by 40% from 0.45 s and the max go up by more than 2 from 881 s); in my test there is about 5% increase in the CPU time.
@rchudasa   Please clarify where the tests were made and also how certain are you that nothing else slowed down the job in the modified reco case.

The only relevant CPU cost increases I see are in 
- ecalDrivenElectronSeeds : \~+1% of the reco job
- allConversions : \~+2% of the reco job
- some other egamma modules with significant self-increase add up to another 1-2% of the reco job.

The (extreme) tails in the CPU processing are due to modules not really related to the changes in ```egamma_lowPt_exclusive``` modifier, primarily from the tracking modules. This means that some tracking activity veto is important for the production setup.


using the same file F9B1A876-3F7D-8449-8437-894248B60877.root and running on 10K events I see slightly different results in 11_0_0_pre5
- default AOD size is 19 kiB/event, compared to 34 kB shown in the reco meeting. 
    - I'm guessing that these 34 kB were obtained with the non-standard event content.  Please check and confirm.
- light-by-light setup with the keep statements is 31 kiB/event, which is closer to the 36 kB shown in the reco meeting.

File: RecoParticleFlow/PFProducer/python/particleFlowBlock_cfi.py (hunk 3 above)
Comment (rchudasa) [reply RecoParticleFlow/PFProducer/python/particleFlowBlock_cfi.py]: could you please suggest how exactly should I do it?

Comment (slava77) [reply RecoParticleFlow/PFProducer/python/particleFlowBlock_cfi.py]: visually, it looks like  replacing 2 with ```_findIndicesByModule("SuperClusterImporter")[0]``` should work,
perhaps more effective is to do
```python
for imp in particleFlowBlock.elementImporters:
  if imp..importerName.value() == "SuperClusterImporter":
    _scImporter = imp
egamma_lowPt_exclusive.toModify(_scImporter, ...
```

**Comment by slava77** [comment]
I ran a reco job for all events in the file and there is at least one (out of 70K) event with about 15% related to the egamma_lowPt configuration updates, compared to 8% increase on average. 

In my test the average time per event with this PR was 1.39 s with a max of 2699 s. Here 
- 98.7% of events completed with time under 4 s/ev and an average in this range of 0.85 s/event. The remaining 2.3% of events with time > 4 s/ev had an average processing time of 43 s/event.
- 99.5% of events completed with time under 10 s/ev and an average in this range of 0.91 s/event. The remaining 0.5% of events with time > 10 s/ev had an average processing time of 103 s/event.

This is a somewhat clear indication of fairly long tails at sub-% level. IMO, these better be filtered out to avoid stuck processing or/and RSS memory over-use which could lead to occasional crashes. 

For this PR the main missing piece now is the event content definition in the code base with the modifier.
Filtering can be addressed separately.

**Comment by rchudasa** [comment]
> about timing: I was not able to reproduce the significant increase in CPU time between the default and the modified recos (you showed the avg go up by 40% from 0.45 s and the max go up by more than 2 from 881 s); in my test there is about 5% increase in the CPU time.
> @rchudasa Please clarify where the tests were made and also how certain are you that nothing else slowed down the job in the modified reco case.
> 
> The only relevant CPU cost increases I see are in
> 
>     * ecalDrivenElectronSeeds : ~+1% of the reco job
> 
>     * allConversions : ~+2% of the reco job
> 
>     * some other egamma modules with significant self-increase add up to another 1-2% of the reco job.
> 
> 
> The (extreme) tails in the CPU processing are due to modules not really related to the changes in `egamma_lowPt_exclusive` modifier, primarily from the tracking modules. This means that some tracking activity veto is important for the production setup.
> 
> using the same file F9B1A876-3F7D-8449-8437-894248B60877.root and running on 10K events I see slightly different results in 11_0_0_pre5
> 
>     * default AOD size is 19 kiB/event, compared to 34 kB shown in the reco meeting.
>       
>       * I'm guessing that these 34 kB were obtained with the non-standard event content.  Please check and confirm.
> 
>     * light-by-light setup with the keep statements is 31 kiB/event, which is closer to the 36 kB shown in the reco meeting.

@slava77 
I re-checked the timing with the default and modified reconstruction for 10K events. I tried to run in CMSSW_10_3_2 and CMSSW_11_0_X_2019-07-29-2300. I used 103X_dataRun2_Prompt_fixEcalADCToGeV_v1 global tag for CMSSW_10_3 and auto:data_run2 for 11_X version.  I am able to get the same results as I presented in my slides for CMSSW_10_3_2 and your results for CMSSW_11_0_X. 

Default reconstruction: 
* CMSSW_10_3_2: Average time: 0.48 s, Max time: 969 s
* CMSSW_11_0_X_2019-07-29-2300: Average time: 0.528 s, Max time: 828 s

Modified reconstruction:
* CMSSW_10_3_2: Average time: 0.64 s, Max time: 2087 s
* CMSSW_11_0_X_2019-07-29-2300: Average time: 0.53 s, Max time: 927 s

and you are right about the event size, default AOD content size is 19kB/event. 
I add the event content definition in code.

**Comment by slava77** [comment]
@cmsbuild please test

**Comment by slava77** [comment]
@rchudasa 
compared to the test command in https://github.com/cms-sw/cmssw/pull/27633#issuecomment-519177827
I'm missing the event content modifications for the following in the latest version of the PR
```
"keep *_interestingGedEgammaIsoESDetId_*_*", "keep *_interestingGedEgammaIsoHCALDetId_*_*", 
"keep *_interestingGedEleIsoDetIdEB_*_*", "keep *_interestingGedEleIsoDetIdEE_*_*",
"keep *_interestingGedGamIsoDetIdEB_*_*","keep *_interestingGedGamIsoDetIdEE_*_*",
"keep EBDigiCollection_ecalDigis__*","keep EEDigiCollection_ecalDigis__*","keep ESDigiCollection_ecalPreshowerDigis__*",
```
Please check and confirm that these are not needed or update the PR to include the missing parts.

Also, compared to the test setup, the following were added
```
     'keep ZDCDataFramesSorted_hcalDigis_*_*',
    'keep ZDCDataFramesSorted_castorDigis_*_*',
    'keep QIE10DataFrameHcalDataFrameContainer_hcalDigis_ZDC_*'
```
ZDC data is small, so it seems OK to add if it's needed.

**Comment by rchudasa** [comment]
> @rchudasa
> compared to the test command in [#27633 (comment)](https://github.com/cms-sw/cmssw/pull/27633#issuecomment-519177827)
> I'm missing the event content modifications for the following in the latest version of the PR
> 
> ```
> "keep *_interestingGedEgammaIsoESDetId_*_*", "keep *_interestingGedEgammaIsoHCALDetId_*_*", 
> "keep *_interestingGedEleIsoDetIdEB_*_*", "keep *_interestingGedEleIsoDetIdEE_*_*",
> "keep *_interestingGedGamIsoDetIdEB_*_*","keep *_interestingGedGamIsoDetIdEE_*_*",
> "keep EBDigiCollection_ecalDigis__*","keep EEDigiCollection_ecalDigis__*","keep ESDigiCollection_ecalPreshowerDigis__*",
> ```
> 
> Please check and confirm that these are not needed or update the PR to include the missing parts.
> 
> Also, compared to the test setup, the following were added
> 
> ```
>      'keep ZDCDataFramesSorted_hcalDigis_*_*',
>     'keep ZDCDataFramesSorted_castorDigis_*_*',
>     'keep QIE10DataFrameHcalDataFrameContainer_hcalDigis_ZDC_*'
> ```
> 
> ZDC data is small, so it seems OK to add if it's needed.

I consulted @Sam-Harper and concluded that we dont need those event contents as they are already present in reducedEgamma rechit collection. We require ZDC event content for the analysis.

**Comment by slava77** [comment]
> I consulted @Sam-Harper and concluded that we dont need those event contents as they are already present in reducedEgamma rechit collection. We require ZDC event content for the analysis.

Do you really mean "reducedEgamma"? This is a module from miniAOD.
In the earlier discussion we concluded that you will not run/need miniAOD.

Perhaps you actually mean other reduced rechit collections available in AOD.
Please clarify/confirm.

**Comment by rchudasa** [comment]
> > I consulted @Sam-Harper and concluded that we dont need those event contents as they are already present in reducedEgamma rechit collection. We require ZDC event content for the analysis.
> 
> Do you really mean "reducedEgamma"? This is a module from miniAOD.
> In the earlier discussion we concluded that you will not run/need miniAOD.
> 
> Perhaps you actually mean other reduced rechit collections available in AOD.
> Please clarify/confirm.

Sorry, for the confusion. We will use AOD content and to be precise, reducedEcalRecHitsEB, reducedEcalRecHitsEE, reducedEcalRecHitsES collections.

**Comment by slava77** [comment]
+1

for  https://github.com/cms-sw/cmssw/pull/27633 fb097ff
- code changes are in line with the PR description and the follow up review
- jenkins tests pass and comparisons with the baseline show no differences (the new process modifier is not exercised)
- the latest updates should not change conclusions made in test in https://github.com/cms-sw/cmssw/pull/27633#issuecomment-519759086 (or in the PR description). Some moderate cleanup of the keep statements  should reduce the event size by \~1 kB/event to just under 30 kB/evt (compared to an increase from 19 to 31 kiB observed in the earlier tests)

**Comment by slava77** [comment]
@fabiocos @kpedro88 
please check this and comment or merge some time soon.
Thank you.

**Comment by kpedro88** [comment]
+1
At some point (once decisions on production are made), a workflow using this modifier should be put into the matrix.

**Comment by kpedro88** [comment]
merge

**Comment by fabiocos** [comment]
+operations

the addition of the modifier is correct, although it is not activated by default in a test workflow
